Sources/Datadog/DatadogCore/Storage/DataBlock.swift

```diff
@@ -132,18 +145,32 @@ internal final class DataBlockReader {
     /// - Throws: `DataBlockError` while reading the input stream.
     /// - Returns: Data bytes from stream.
     private func read(length: Int) throws -> Data {
-        var bytes = [UInt8](repeating: 0, count: length)
-        let count = stream.read(&bytes, maxLength: length)
+        guard length > 0 else {
+            return Data()
+        }
+
+        // Load from stream to data without unnecessary copies
+        var data = Data(repeating: 0, count: length)
+        let count = try data.withUnsafeMutableBytes { (bytes: UnsafeMutableRawBufferPointer) in
+            guard let buffer = bytes.assumingMemoryBound(to: UInt8.self).baseAddress else {
+                throw DataBlockError.dataAllocationFailure
+            }
+            return stream.read(buffer, maxLength: length)
+        }
```

Comment:
So before there was 3 copies of the data:
1. `file.read()` read the entire contents into a `Data`.
2. `bytes` the `[UInt8]` buffer
3. The `Data` that was initialized with `bytes`. This is a copy.

So 2 was very short lived and it's possible that the swift optimizer removes that copy, but I don't know for sure.

1 was solved by using the `InputStream` api that loads directly from a file, instead of from `Data`. 2 and 3 were then merged using `Data(repeating: 0, count: length)` and then reading directly to that `Data`.

There is also probably an optimization where you use an [uninitialized `Data`](https://developer.apple.com/documentation/foundation/data/1780072-init), but I was having some trouble and didn't want to spend too much time on it. I just noticed [this api](https://developer.apple.com/documentation/foundation/data/2201934-init) which is probably better than using `repeating: 0` as well.

Reply:
Thanks for explanation 👍 - now it makes sense given this broader picture. I was wrong by thinking that "copies" relate to local scope in this method.